File: internal/forms/registry.go
package forms

import (
	"fmt"
	"strings"
	"sync"
)

// FormConstructor is a function that returns a FormDef. Form packages
// call RegisterForm in their init() functions to add constructors to
// the global auto-registration list.
type FormConstructor func() *FormDef

var (
	autoFormsMu sync.Mutex
	autoForms   []FormConstructor
)

// RegisterForm adds a FormDef constructor to the global auto-registration
// list. It is intended to be called from init() functions in form packages.
func RegisterForm(fn FormConstructor) {
	autoFormsMu.Lock()
	defer autoFormsMu.Unlock()
	autoForms = append(autoForms, fn)
}

// AutoRegisteredForms returns FormDefs from all constructors that have
// been registered via RegisterForm. Each constructor is called once.
func AutoRegisteredForms() []*FormDef {
	autoFormsMu.Lock()
	defer autoFormsMu.Unlock()
	result := make([]*FormDef, 0, len(autoForms))
	for _, fn := range autoForms {
		result = append(result, fn())
	}
	return result
}

// NewRegistryFromAutoForms creates a Registry pre-populated with every
// form that was auto-registered via init() + RegisterForm.
func NewRegistryFromAutoForms() *Registry {
	reg := NewRegistry()
	for _, form := range AutoRegisteredForms() {
		reg.Register(form)
	}
	return reg
}

// Registry holds all registered forms and provides lookup.
type Registry struct {
	forms map[FormID]*FormDef
}

// NewRegistry creates an empty Registry.
func NewRegistry() *Registry {
	return &Registry{
		forms: make(map[FormID]*FormDef),
	}
}

// Register adds a form definition to the registry.
func (r *Registry) Register(form *FormDef) {
	r.forms[form.ID] = form
}

// Get returns the form with the given ID, or false if not found.
func (r *Registry) Get(formID FormID) (*FormDef, bool) {
	f, ok := r.forms[formID]
	return f, ok
}

// GetField looks up a field by its fully qualified key ("form_id:line").
// Returns the parent form and field definition, or an error if not found.
func (r *Registry) GetField(key string) (*FormDef, *FieldDef, error) {
	parts := strings.SplitN(key, ":", 2)
	if len(parts) != 2 {
		return nil, nil, fmt.Errorf("invalid field key %q: expected form_id:line", key)
	}
	formID, line := FormID(parts[0]), parts[1]

	form, ok := r.forms[formID]
	if !ok {
		return nil, nil, fmt.Errorf("form %q not found", formID)
	}

	field := form.FieldByLine(line)
	if field == nil {
		return nil, nil, fmt.Errorf("field %q not found in form %q", line, formID)
	}
	return form, field, nil
}

// AllForms returns all registered form definitions.
func (r *Registry) AllForms() []*FormDef {
	result := make([]*FormDef, 0, len(r.forms))
	for _, f := range r.forms {
		result = append(result, f)
	}
	return result
}

// ValidateFieldDefs checks all registered forms for common errors.
func (r *Registry) ValidateFieldDefs() []error {
	var errs []error
	for _, form := range r.AllForms() {
		for _, field := range form.Fields {
			if field.Type == UserInput && field.Compute != nil {
				errs = append(errs, fmt.Errorf("form %s field %s: UserInput should not have Compute", form.ID, field.Line))
			}
		}
	}
	return errs
}

// ValidateFederalRefs checks that all FederalRef fields reference forms with
// Jurisdiction == Federal. Returns a list of errors for any violations.
func (r *Registry) ValidateFederalRefs() []error {
	var errs []error
	for _, form := range r.AllForms() {
		for _, field := range form.Fields {
			if field.Type != FederalRef {
				continue
			}
			for _, dep := range field.DependsOn {
				parts := strings.SplitN(dep, ":", 2)
				if len(parts) != 2 {
					errs = append(errs, fmt.Errorf("form %s field %s: FederalRef dependency %q has invalid format", form.ID, field.Line, dep))
					continue
				}
				depFormID := FormID(parts[0])
				depForm, ok := r.forms[depFormID]
				if !ok {
					// The referenced form may not be registered (e.g., input forms).
					// Only validate forms that are present in the registry.
					continue
				}
				if depForm.Jurisdiction != Federal {
					errs = append(errs, fmt.Errorf("form %s field %s: FederalRef dependency %q references non-federal form %s (jurisdiction %d)", form.ID, field.Line, dep, depFormID, depForm.Jurisdiction))
				}
			}
		}
	}
	return errs
}

File: api/src/domain/registry.rs
use std::collections::HashMap;

use crate::domain::field::{FieldDef, FieldType, FormDef, Jurisdiction};

/// Registry holds all registered forms and provides lookup.
pub struct Registry {
    forms: HashMap<String, FormDef>,
}

impl Registry {
    /// Creates an empty Registry.
    pub fn new() -> Self {
        Self {
            forms: HashMap::new(),
        }
    }

    /// Adds a form definition to the registry.
    pub fn register(&mut self, form: FormDef) {
        self.forms.insert(form.id.clone(), form);
    }

    /// Returns the form with the given ID, or None if not found.
    pub fn get(&self, form_id: &str) -> Option<&FormDef> {
        self.forms.get(form_id)
    }

    /// Looks up a field by its fully qualified key ("form_id:line").
    /// Returns the parent form and field definition, or an error if not found.
    pub fn get_field(&self, key: &str) -> Result<(&FormDef, &FieldDef), String> {
        let parts: Vec<&str> = key.splitn(2, ':').collect();
        if parts.len() != 2 {
            return Err(format!(
                "invalid field key {:?}: expected form_id:line",
                key
            ));
        }
        let (form_id, line) = (parts[0], parts[1]);

        let form = self
            .forms
            .get(form_id)
            .ok_or_else(|| format!("form {:?} not found", form_id))?;

        let field = form
            .field_by_line(line)
            .ok_or_else(|| format!("field {:?} not found in form {:?}", line, form_id))?;

        Ok((form, field))
    }

    /// Returns all registered form definitions.
    pub fn all_forms(&self) -> Vec<&FormDef> {
        self.forms.values().collect()
    }

    /// Checks all registered forms for common errors.
    /// Returns a list of error messages.
    pub fn validate_field_defs(&self) -> Vec<String> {
        let mut errs = Vec::new();
        for form in self.forms.values() {
            for field in &form.fields {
                if field.field_type == FieldType::UserInput && field.compute.is_some() {
                    errs.push(format!(
                        "form {} field {}: UserInput should not have Compute",
                        form.id, field.line
                    ));
                }
            }
        }
        errs
    }

    /// Checks that all FederalRef fields reference forms with Jurisdiction::Federal.
    pub fn validate_federal_refs(&self) -> Vec<String> {
        let mut errs = Vec::new();
        for form in self.forms.values() {
            for field in &form.fields {
                if field.field_type != FieldType::FederalRef {
                    continue;
                }
                for dep in &field.depends_on {
                    let parts: Vec<&str> = dep.splitn(2, ':').collect();
                    if parts.len() != 2 {
                        errs.push(format!(
                            "form {} field {}: FederalRef dependency {:?} has invalid format",
                            form.id, field.line, dep
                        ));
                        continue;
                    }
                    let dep_form_id = parts[0];
                    if let Some(dep_form) = self.forms.get(dep_form_id) {
                        if dep_form.jurisdiction != Jurisdiction::Federal {
                            errs.push(format!(
                                "form {} field {}: FederalRef dependency {:?} references non-federal form {} (jurisdiction {:?})",
                                form.id, field.line, dep, dep_form_id, dep_form.jurisdiction
                            ));
                        }
                    }
                    // If the referenced form is not registered, skip validation
                }
            }
        }
        errs
    }
}

impl Default for Registry {
    fn default() -> Self {
        Self::new()
    }
}
